### PR TITLE
HBASE-24898 Addendum to deal with Daylight Saving Time

### PR DESCRIPTION
**Pull request by bsglz**

### CODE REVIEW

**Comment by bsglz** [comment]
> Are we sure this is enough? No timezone other than Daylight is causing this edge case miss?

Yeah, for special datetime such as 2020-08-20 11:52:41 and 2020-08-20 15:04:00, i think so, but not sure if there are different calculation for other dates.
Considering this case is target to test if the EnvironmentEdge take effect correctly in CurrentHourProvider, maybe it is enough?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/CurrentHourProvider.java [hunk 1]
```diff
@@ -37,7 +38,8 @@
     }
   }
 
-  private static Tick nextTick() {
+  @VisibleForTesting
+  protected static Tick nextTick() {
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/CurrentHourProvider.java:42]: Just remove the protected qualifier? We just want it to be package private right?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/CurrentHourProvider.java [hunk 2]
```diff
@@ -52,7 +54,8 @@ private static void moveToNextHour(Calendar calendar) {
     calendar.set(Calendar.MILLISECOND, 0);
   }
 
-  private static volatile Tick tick = nextTick();
+  @VisibleForTesting
+  protected static volatile Tick tick = nextTick();
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/CurrentHourProvider.java:58]: Ditto.

**Review by Apache9 — CHANGES_REQUESTED** [review]
I think we could let someone in Bay Area to test if the UT is OK for them?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/CurrentHourProvider.java (hunk 1 above)
Comment (bsglz) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/CurrentHourProvider.java:42]: Will fix later, thanks.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/CurrentHourProvider.java (hunk 2 above)
Comment (bsglz) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/CurrentHourProvider.java:58]: Will fix later, thanks.

**Comment by bsglz** [comment]
> I think we could let someone in Bay Area to test if the UT is OK for them?

Yeah, or we can change the timezone of os.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m  4s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 45s |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 15s |  master passed  |
| +1 :green_heart: |  spotbugs  |   2m 12s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 46s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 11s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  11m 21s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.1.  |
| +1 :green_heart: |  spotbugs  |   2m  8s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 16s |  The patch does not generate ASF License warnings.  |
|  |   |  34m 12s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.12 Server=19.03.12 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2310 |
| Optional Tests | dupname asflicense spotbugs hadoopcheck hbaseanti checkstyle |
| uname | Linux f3692b217704 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / ebe321a99b |
| Max. process+thread count | 94 (vs. ulimit of 12500) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/console |
| versions | git=2.17.1 maven=(cecedd343002696d0abb50b32b541b8a6ba2883f) spotbugs=3.1.12 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 25s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 49s |  master passed  |
| +1 :green_heart: |  compile  |   0m 56s |  master passed  |
| +1 :green_heart: |  shadedjars  |   6m 30s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 39s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 31s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 56s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 56s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   6m 32s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 37s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 139m 17s |  hbase-server in the patch passed.  |
|  |   | 165m 23s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.12 Server=19.03.12 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/artifact/yetus-jdk8-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2310 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux 78886966c042 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / ebe321a99b |
| Default Java | 1.8.0_232 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/testReport/ |
| Max. process+thread count | 4795 (vs. ulimit of 12500) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/console |
| versions | git=2.17.1 maven=(cecedd343002696d0abb50b32b541b8a6ba2883f) |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 29s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 27s |  master passed  |
| +1 :green_heart: |  compile  |   1m  6s |  master passed  |
| +1 :green_heart: |  shadedjars  |   8m  5s |  branch has no errors when building our shaded downstream artifacts.  |
| -0 :warning: |  javadoc  |   0m 45s |  hbase-server in master failed.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m  6s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   7m 15s |  patch has no errors when building our shaded downstream artifacts.  |
| -0 :warning: |  javadoc  |   0m 41s |  hbase-server in the patch failed.  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 148m 56s |  hbase-server in the patch failed.  |
|  |   | 178m 59s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.12 Server=19.03.12 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/artifact/yetus-jdk11-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2310 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux 63dc4d8f9d87 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / ebe321a99b |
| Default Java | 2020-01-14 |
| javadoc | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/artifact/yetus-jdk11-hadoop3-check/output/branch-javadoc-hbase-server.txt |
| javadoc | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/artifact/yetus-jdk11-hadoop3-check/output/patch-javadoc-hbase-server.txt |
| unit | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/artifact/yetus-jdk11-hadoop3-check/output/patch-unit-hbase-server.txt |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/testReport/ |
| Max. process+thread count | 3781 (vs. ulimit of 12500) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/1/console |
| versions | git=2.17.1 maven=(cecedd343002696d0abb50b32b541b8a6ba2883f) |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m  7s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   6m 31s |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 46s |  master passed  |
| +1 :green_heart: |  spotbugs  |   3m 19s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   6m 14s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 49s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  16m 25s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.1.  |
| +1 :green_heart: |  spotbugs  |   2m 51s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 17s |  The patch does not generate ASF License warnings.  |
|  |   |  52m  8s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.12 Server=19.03.12 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2310 |
| Optional Tests | dupname asflicense spotbugs hadoopcheck hbaseanti checkstyle |
| uname | Linux 541499cc18c1 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 01cf60067c |
| Max. process+thread count | 84 (vs. ulimit of 12500) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/2/console |
| versions | git=2.17.1 maven=(cecedd343002696d0abb50b32b541b8a6ba2883f) spotbugs=3.1.12 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Review by Apache9 — APPROVED** [review]

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 39s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 17s |  master passed  |
| +1 :green_heart: |  compile  |   0m 58s |  master passed  |
| +1 :green_heart: |  shadedjars  |   8m 40s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 43s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 41s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  0s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  0s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   6m 46s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 38s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 141m 44s |  hbase-server in the patch passed.  |
|  |   | 171m 16s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.12 Server=19.03.12 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/2/artifact/yetus-jdk8-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2310 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux 6a3fdacbf958 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 01cf60067c |
| Default Java | 1.8.0_232 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/2/testReport/ |
| Max. process+thread count | 4124 (vs. ulimit of 12500) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/2/console |
| versions | git=2.17.1 maven=(cecedd343002696d0abb50b32b541b8a6ba2883f) |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 47s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 40s |  master passed  |
| +1 :green_heart: |  compile  |   1m 17s |  master passed  |
| +1 :green_heart: |  shadedjars  |   7m 55s |  branch has no errors when building our shaded downstream artifacts.  |
| -0 :warning: |  javadoc  |   0m 45s |  hbase-server in master failed.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 48s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 22s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 22s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   7m 56s |  patch has no errors when building our shaded downstream artifacts.  |
| -0 :warning: |  javadoc  |   0m 52s |  hbase-server in the patch failed.  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 143m 13s |  hbase-server in the patch passed.  |
|  |   | 175m 36s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.12 Server=19.03.12 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/2/artifact/yetus-jdk11-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2310 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux 09b06d0f9c46 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 01cf60067c |
| Default Java | 2020-01-14 |
| javadoc | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/2/artifact/yetus-jdk11-hadoop3-check/output/branch-javadoc-hbase-server.txt |
| javadoc | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/2/artifact/yetus-jdk11-hadoop3-check/output/patch-javadoc-hbase-server.txt |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/2/testReport/ |
| Max. process+thread count | 4319 (vs. ulimit of 12500) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2310/2/console |
| versions | git=2.17.1 maven=(cecedd343002696d0abb50b32b541b8a6ba2883f) |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Review by virajjasani — APPROVED** [review]
+1
